src/mono/mono/mini/intrinsics.c

```diff
@@ -556,7 +556,10 @@ MONO_RESTORE_WARNING
 		t = mini_get_underlying_type (t);
 		if (cfg->gshared && t != ctx->method_inst->type_argv [0] && MONO_TYPE_ISSTRUCT (t) && mono_class_check_context_used (mono_class_from_mono_type_internal (t)))
 			cfg->prefer_instances = TRUE;
-		return mini_emit_memory_load (cfg, t, args [0], 0, MONO_INST_UNALIGNED);
+			
+		int align;
+		mono_type_size (t, &align);
+		return mini_emit_memory_load (cfg, t, args [0], 0, ((guintptr)&(args [0]->dreg) % align == 0) ? 0 : MONO_INST_UNALIGNED);
```

Comment:
I don't think `((guintptr)&(args [0]->dreg) % align == 0)` does the right thing. That tests address of a member in a structure, so it's essentially a constant expression at type of emit.

You can always emit normal load on x86/x64 because the processors do support unaligned reads internally (with a performance penalty). For everything else it would need to emit a runtime check which is possibly not any better than unrolled memcpy.

Reply:
Ah, I see. I thought that the `agrs [0]->dreg` points to the actual address we are trying to read from.